ckanext/security/middleware.py

```diff
@@ -68,7 +77,14 @@ def check_cookie(self, request):
         if token is None:
             # Just in case this is set by an AJAX request
             token = request.cookies.get('X-CSRFToken', None)
-        return compare_digest(token, self.cache.get(self.session.id))
+        
+        csrf_token = self.cache.get(self.session.id)
+
+        if csrf_token == None:
+            log.warning('Could not find a csrf token for session id: {}\n{}'.format(self.session.id, request))
+            return False
```

Comment:
This was being caused by the `if 'text/html' in resp.headers['Content-type']:` conditional in the `__call__` method.

@anotheredward I'm a little shaky on the entire behaviour of this module, so would be keen to get your input on how we could better structure this middleware, so that we are only updating redis when necessary.

Reply:
Ok, but note that without this the CSRF tokens now have a race condition, as every request changes the token, so the token that is currently set in the user's cookie may not be what the server has.

This was the original bug that made form submissions not work most of the time, especially when there were any slow responses...

It's possible that this was fixed in production by serving static files with nginx rather than the application, but it certainly breaks our development environment.

Reply:
I re-added this conditional and tested it again, I was not able to reproduce the issue with a null csrf_token.
I believe it is possible I was messing up my test environment by having two browsers pointing at the site at once.

I've added a todo to re-add the conditional before merging this feature.

Reply:
@anotheredward, @ebuckley I have the same issue - every request changes the token in __call__ method, so the csrf_token is always None. Have you find the solution for this?

Reply:
In the issue that we saw the csrf_token was no longer being set for API requests, which is why @ebuckley made the above change.

I had previously made a change to only set the csrf_token when the content-type was text/html to avoid resetting the csrf_token every request, but this is what caused the issue and locked users out of api access.

I'm not sure why your csrf_token is always None, it might be another configuration issue as this is currently working on our instance of ckan.